x/datapool/types/message_pool.go

```diff
@@ -0,0 +1,168 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var _ sdk.Msg = &MsgRegisterDataValidator{}
+
+func NewMsgRegisterDataValidator(address string, dataValidator *DataValidator) *MsgRegisterDataValidator {
+	return &MsgRegisterDataValidator{
+		Address:         address,
+		ValidatorDetail: dataValidator,
+	}
+}
+
+func (msg *MsgRegisterDataValidator) Route() string {
+	return RouterKey
+}
+
+func (msg *MsgRegisterDataValidator) Type() string {
+	return "RegisterDataValidator"
+}
+
+func (msg *MsgRegisterDataValidator) ValidateBasic() error {
```

Comment:
It would be good to include an empty value check for `endpoints`.

Reply:
I'll modify this function in next PR.